src/rust/engine/remote_provider/remote_provider_reapi/src/byte_store.rs

```diff
@@ -390,6 +444,139 @@ impl ByteStoreProvider for Provider {
+            - self
+                .instance_name
+                .as_ref()
+                .cloned()
+                .unwrap_or_default()
+                .len()
+            - 10;
+
+        let mut chunks: Vec<Vec<Digest>> = Vec::new();
+        let mut current_chunk: Vec<Digest> = Vec::new();
+        let mut current_size = 0;
+
+        for digest in digests.iter() {
+            let message_size = digest.size_bytes + RPC_RESPONSE_PER_ITEM_SIZE;
+            if current_size + message_size >= max_batch_total_size_bytes {
+                chunks.push(std::mem::take(&mut current_chunk));
+                current_size = 0;
+            }
+            current_chunk.push(*digest);
+            current_size += message_size;
+        }
+
+        if !current_chunk.is_empty() {
+            chunks.push(current_chunk);
+        }
+
+        let client = self.cas_client.as_ref();
+        let destination = Arc::new(Mutex::new(destination));
+
+        let chunk_futures = chunks
+            .clone()
+            .into_iter()
+            .map(|chunk| {
+                let digests = chunk
+                    .iter()
+                    .map(|d| d.into())
+                    .collect::<Vec<remexec::Digest>>();
+                let request = remexec::BatchReadBlobsRequest {
+                    instance_name: self.instance_name.as_ref().cloned().unwrap_or_default(),
+                    digests: digests,
+                    acceptable_compressors: vec![],
+                };
+
+                let client = client.clone();
+                let destination = destination.clone();
+
+                retry_call(
+                    client,
+                    move |mut client, _| {
+                        let request = request.clone();
+                        let chunk = chunk.clone();
+                        let destination = destination.clone();
+
+                        async move {
+                            let response = client.batch_read_blobs(request).await?;
+                            let responses =
+                                Provider::validate_batch_response(chunk, response.into_inner())
+                                    .map_err(Status::unknown)?;
+
+                            if let Some(err) = responses.values().find_map(|res| res.as_ref().err())
+                            {
+                                return Err(err.clone());
+                            }
+
+                            let to_write = responses
+                                .into_iter()
+                                .map(|(d, res)| (d, res.unwrap()))
+                                .collect::<Vec<_>>();
+
+                            let digests = to_write.iter().map(|(d, _)| (*d)).collect::<Vec<_>>();
+
+                            destination
+                                .lock()
+                                .await
+                                .write(to_write)
```

Comment:
I need to write the bytes out to their destination in this future before the `join_all` below so they're not all in memory until then